<div style="text-align: center; background-color: #0A6EBD; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Lab 01 - Data visualization with Python @ FIT-HCMUS, VNU-HCM 📌
</div>

<div style="text-align: center; background-color: #b1d1ff; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Stage 02 - Data exploration 📌
</div>

# Import các thư viện cần thiết

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Khám phá dữ liệu

### Đọc dữ liệu từ file

In [2]:
raw_df = None
raw_df = pd.read_csv('../data/raw/used_cars_data.csv')

In [3]:
raw_df

S.No.                                               Name    Location  \
0         0                             Maruti Wagon R LXI CNG      Mumbai   
1         1                   Hyundai Creta 1.6 CRDi SX Option        Pune   
2         2                                       Honda Jazz V     Chennai   
3         3                                  Maruti Ertiga VDI     Chennai   
4         4                    Audi A4 New 2.0 TDI Multitronic  Coimbatore   
...     ...                                                ...         ...   
7248   7248                  Volkswagen Vento Diesel Trendline   Hyderabad   
7249   7249                             Volkswagen Polo GT TSI      Mumbai   
7250   7250                             Nissan Micra Diesel XV     Kolkata   
7251   7251                             Volkswagen Polo GT TSI        Pune   
7252   7252  Mercedes-Benz E-Class 2009-2013 E 220 CDI Avan...       Kochi   

      Year  Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
0     2010              72000       CNG       Manual      First  26.6 km/kg   
1     2015              41000    Diesel       Manual      First  19.67 kmpl   
2     2011              46000    Petrol       Manual      First   18.2 kmpl   
3     2012              87000    Diesel       Manual      First  20.77 kmpl   
4     2013              40670    Diesel    Automatic     Second   15.2 kmpl   
...    ...                ...       ...          ...        ...         ...   
7248  2011              89411    Diesel       Manual      First  20.54 kmpl   
7249  2015              59000    Petrol    Automatic      First  17.21 kmpl   
7250  2012              28000    Diesel       Manual      First  23.08 kmpl   
7251  2013              52262    Petrol    Automatic      Third   17.2 kmpl   
7252  2014              72443    Diesel    Automatic      First   10.0 kmpl   

       Engine      Power  Seats  New_Price  Price  
0      998 CC  58.16 bhp    5.0        NaN   1.75  
1     1582 CC  126.2 bhp    5.0        NaN  12.50  
2     1199 CC   88.7 bhp    5.0  8.61 Lakh   4.50  
3     1248 CC  88.76 bhp    7.0        NaN   6.00  
4     1968 CC  140.8 bhp    5.0        NaN  17.74  
...       ...        ...    ...        ...    ...  
7248  1598 CC  103.6 bhp    5.0        NaN    NaN  
7249  1197 CC  103.6 bhp    5.0        NaN    NaN  
7250  1461 CC   63.1 bhp    5.0        NaN    NaN  
7251  1197 CC  103.6 bhp    5.0        NaN    NaN  
7252  2148 CC    170 bhp    5.0        NaN    NaN  

[7253 rows x 14 columns]

### Xóa cột thứ tự không cần thiết

In [4]:
raw_df = raw_df.drop('S.No.', axis=1)

### Tạo thêm một cột Company name

In [5]:
raw_df['Company name'] = raw_df['Name'].apply(lambda x: ' '.join(x.split()[:2]) if x.split()[0] in ['Mini', 'Land'] else x.split()[0])

### Dữ liệu có bao nhiêu cột và bao nhiêu dòng?

In [6]:
shape = raw_df.shape
shape

(7253, 14)

### Ý nghĩa của mỗi dòng là gì?

Mỗi dòng chứa các thông tin của mỗi chiếc xe đã qua sử dụng và được đăng bán lại như tên xe, địa điểm, năm đăng ký bán lại, các thuộc tính của xe, giá bán chủ yếu là các thành phố của Ấn Độ.

### Dữ liệu có dòng nào bị trùng lặp không?

In [7]:
num_duplicated_rows = None
num_duplicated_rows = raw_df.duplicated().sum()
#TEST
if num_duplicated_rows == 0:
    print(f"Dữ liệu không có dòng nào trùng lặp!")
else:
    print(f"Dữ liệu có {num_duplicated_rows} dòng bị trùng lặp "+ ". Vui lòng xóa các dòng đó!")

Dữ liệu có 1 dòng bị trùng lặp . Vui lòng xóa các dòng đó!


#### Xóa các dòng bị trùng lắp dữ liệu

In [8]:

raw_df = raw_df.drop_duplicates()

#### Kiểm tra lại

In [9]:
#TEST
num_duplicated_rows = raw_df.duplicated().sum()
if num_duplicated_rows == 0:
    print(f"Dữ liệu không có dòng nào trùng lặp!")
else:
    print(f"Dữ liệu có {num_duplicated_rows} dòng bị trùng lặp "+ ". Vui lòng xóa các dòng đó!")

Dữ liệu không có dòng nào trùng lặp!


### Ý nghĩa của mỗi cột là gì?

- Name: Tên của ô tô.
- Location: Nơi đăng ký bán lại của ô tô đó.
- Year: Năm đăng ký bán lại ô tô đó.
- Kilometers_Driven: quãng đường(số km) mà xe đã đi.
- Fuel_Type: Loại nhiên liệu xe sử dụng.
- Transmission: Loại hộp số của xe.
- Owner_Type: Người dùng đời thứ mấy của xe.
- Mileage: Quãng đường mà xe đi được trong một đơn vị nhiên liệu như kmpl(kilometers per liter) hoặc km/kg.
- Engine: Dung tích xi-lanh của động cơ xe sử dụng đơn vị là cc.
- Power: Công xuất tối đa của xe sử dụng đơn vị là bhp (brake horsepower - đơn vị đo lường để mô tả lượng điện năng mà động cơ ô tô hoặc động cơ điện tạo ra để di chuyển nó).
- Seats: Số chỗ ngồi của xe.
- New_Price: Giá bán mới cập nhật của xe được sử dụng đơn vị Lakh (là một loại ngôn ngữ chỉ số lượng lớn trong ngôn ngữ Ấn Độ, 1 Lakh = 100000).
- Price: Giá bán ban đầu của xe được sử dụng đơn vị Lakh.

### Mỗi cột có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu không phù hợp cho việc xử lí dữ liệu không?

In [10]:

raw_df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 7252 entries, 0 to 7252
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               7252 non-null   object 
 1   Location           7252 non-null   object 
 2   Year               7252 non-null   int64  
 3   Kilometers_Driven  7252 non-null   int64  
 4   Fuel_Type          7252 non-null   object 
 5   Transmission       7252 non-null   object 
 6   Owner_Type         7252 non-null   object 
 7   Mileage            7250 non-null   object 
 8   Engine             7206 non-null   object 
 9   Power              7206 non-null   object 
 10  Seats              7199 non-null   float64
 11  New_Price          1006 non-null   object 
 12  Price              6019 non-null   float64
 13  Company name       7252 non-null   object 
dtypes: float64(2), int64(2), object(10)
memory usage: 849.8+ KB


- Ở cột Engine, Power và New_Price chủ yếu là kiểu float nhưng do có thêm kí tự đơn vị phía sau nên đang bị thuộc về loại categorical.
- Cột Year nên thuộc loại categorical.

In [11]:
raw_df['Year'] = raw_df['Year'].astype('string')

### Xử lý datatype cho 3 cột Engine, Power và New_Price

In [12]:
raw_df['Engine'] = raw_df['Engine'].str.replace(' CC', '')
raw_df['Power'] = raw_df['Power'].str.replace(' bhp', '')
raw_df['New_Price'] = raw_df['New_Price'].str.replace(' Lakh', '')


In [13]:
# Mileage = ?
raw_df['Engine'] = pd.to_numeric(raw_df['Engine'], errors='coerce')
raw_df['Power'] = pd.to_numeric(raw_df['Power'], errors='coerce')
raw_df['New_Price'] = pd.to_numeric(raw_df['New_Price'], errors='coerce')

### Xử lý cột Mileage
Vì ở cột này có loại đơn vị khác nhau, nên chúng ta sẽ chuyển đổi đơn vị km/kg sang kmpl

In [14]:
Correct_Mileage= []
for i in raw_df.Mileage:
    if str(i).endswith('km/kg'):
        i = i[:-6]
        i = float(i)*1.40
        Correct_Mileage.append(float(i))
    elif str(i).endswith('kmpl'):
        i = i[:-6]
        #print(i)
        Correct_Mileage.append(float(i))
    else:
        Correct_Mileage.append(float(i))
raw_df['Mileage']=Correct_Mileage
raw_df['Mileage'] = pd.to_numeric(raw_df['Mileage'], errors='coerce')


### Giá trị của các cột có kiểu numeric phân bố như thế nào?
- Phần trăm các giá trị bị thiếu là bao nhiêu?
- Giá trị lớn nhất? Giá trị nhỏ nhất? Có gì bất thường không?

Đối với các cột có kiểu numeric, chúng ta sẽ tính:
- Phần trăm (từ 0 tới 100) của giá trị bị thiếu
- Giá trị nhỏ nhất
- Tứ phân vị 25
- Tứ phân vị 50
- Tứ phân vị 75
- Giá trị lớn nhất

Lưu kết quả vào DataFrame `num_col_info_df`, trong đó:
- Tên của các cột là tên của các cột có kiểu numeric trong `df`
- Tên của các dòng: "missing_ratio", "min", "lower_quartile", "median", "upper_quartile", "max"

In [15]:
num_col_info_df = None

num_col_info_df = raw_df.select_dtypes(include=np.number).copy()

def missing_ratio(column):
    return ((column.isnull().sum() / column.shape[0]) * 100).round(1)

def lower_quartile(column):
    return (column.quantile(0.25)).round(1)

def median(column):
    return (column.median())

def upper_quartile(column):
    return (column.quantile(0.75)).round(1)

# Làm tròn giá trị đến 1 chữ số thập phân
num_col_info_df = num_col_info_df.round(1)

num_col_info_df = num_col_info_df.agg([missing_ratio, "min", lower_quartile, median, upper_quartile, "max"])

In [16]:
num_col_info_df

Kilometers_Driven  Mileage  Engine  Power  Seats  New_Price  \
missing_ratio                 0.0      0.0     0.6    2.4    0.7       86.4   
min                         171.0      0.0    72.0   34.2    0.0        3.9   
lower_quartile            34000.0     15.0  1198.0   75.0    5.0        7.9   
median                    53429.0     18.0  1493.0   94.0    5.0       11.5   
upper_quartile            73000.0     21.0  1968.0  138.1    5.0       24.0   
max                     6500000.0     47.0  5998.0  616.0   10.0       99.9   

                Price  
missing_ratio    17.0  
min               0.4  
lower_quartile    3.5  
median            5.6  
upper_quartile   10.0  
max             160.0

#### Có điều bất thường là: 
- New_Price có tỉ lệ mất mát rất cao lên đến 86.4%
- Giá trị max của Kilometers_Driven là nhiễu.

=> Vậy chúng ta xử lý bằng cách xóa hoàn toàn cột New_Price.

### Xóa giá trị max đang là nhiễu

In [17]:
raw_df = raw_df[raw_df['Kilometers_Driven'] != 6500000]

#### Xóa cột có tỉ lệ giá trị bị thiếu lớn hơn 75%

In [18]:
def drop_missing_features(df: pd.DataFrame, missing_lst: dict = dict(num_col_info_df.iloc[0]), threshold: float =75.0) -> pd.DataFrame:
    for col, missing_rate in missing_lst.items():
        if missing_rate > threshold:
            del df[col]
    return df

In [19]:
raw_df = drop_missing_features(raw_df)

### Xử lý những dòng bị mất giá trị bằng cách tìm các dòng khác có cùng tên xe và thay thế giá trị của dòng khác vào dòng bị mất

In [20]:

numeric_col_process = ['Mileage', 'Engine', 'Power', 'Seats', 'Price']

In [21]:
for col in numeric_col_process:
    rows_with_nan = raw_df[raw_df[col].isna()]

# Duyệt qua từng dòng có NaN trong col
for index, row in rows_with_nan.iterrows():
    # Tìm dòng có 'Name' trùng với dòng hiện tại
    matched_rows = raw_df[(raw_df['Name'] == row['Name']) & (~raw_df[col].isna())]
    if not matched_rows.empty:
        # Nếu tìm thấy ít nhất một dòng trùng khớp
        matched_row = matched_rows.iloc[0]
        # Thay thế giá trị NaN trong col bằng giá trị tương ứng trong dòng trùng khớp
        raw_df.at[index, col] = matched_row[col]

#### Các dòng còn lại có thể sẽ không có giá nên nhóm sẽ thay bằng median để ít làm ảnh hưởng đến dữ liệu nhất có thể

In [22]:
for col in numeric_col_process:
    raw_df[col] = raw_df[col].fillna(raw_df[col].median())


### Giá trị của các cột có kiểu categorical phân bố như thế nào?
- Phần trăm giá trị bị thiếu là bao nhiêu?
- Có bao nhiêu giá trị phân biệt? Hiển thị một vài giá trị
- Có gì bất thường không?

Đối với các cột có kiểu categorical, chúng ta sẽ tính:
- Phần trăm (từ 0 tới 100) của giá trị bị thiếu
- Số lượng giá trị (giá trị phân biệt, không xét các giá trị bị thiếu)
Lưu kết quả vào DataFrame num_col_info_df, trong đó:
- Tên của các cột là tên của các cột có kiểu không phải numeric trong `df`
- Tên của các dòng: "missing_ratio", "num_values"

In [23]:
cat_col_info_df = raw_df.select_dtypes(exclude=[np.number])

def missing_ratio(column):
    return ((column.isnull().sum() / column.shape[0]) * 100).round(1)

# Hàm tính số lượng giá trị
def num_values(column):
    return column.nunique()

# Hàm tính tỷ lệ của từng giá trị
def value_ratios(column):
    value_counts = column.value_counts() #Đếm số lượng của mỗi loại value trong 1 cột
    non_missing_count = value_counts.sum() #Tổng số lượng của tất cả value trong 1 cột
    ratios = (value_counts / non_missing_count * 100).round(1) #Lưu tỉ lệ vào Series
    ratios_dict = ratios.to_dict()
    sorted_ratios_dict = dict(sorted(ratios_dict.items(), key=lambda item: item[1], reverse=True))
    return sorted_ratios_dict

cat_col_info_df = cat_col_info_df.agg([missing_ratio, num_values, value_ratios])
cat_col_info_df

Name  \
missing_ratio                                                0.0   
num_values                                                  2041   
value_ratios   {'Mahindra XUV500 W8 2WD': 0.8, 'Maruti Swift ...   

                                                        Location  \
missing_ratio                                                0.0   
num_values                                                    11   
value_ratios   {'Mumbai': 13.1, 'Hyderabad': 12.1, 'Coimbator...   

                                                            Year  \
missing_ratio                                                0.0   
num_values                                                    23   
value_ratios   {'2015': 12.8, '2014': 12.8, '2016': 12.2, '20...   

                                                       Fuel_Type  \
missing_ratio                                                0.0   
num_values                                                     5   
value_ratios   {'Diesel': 53.1, 'Petrol': 45.8, 'CNG': 0.9, '...   

                                      Transmission  \
missing_ratio                                  0.0   
num_values                                       2   
value_ratios   {'Manual': 71.8, 'Automatic': 28.2}   

                                                      Owner_Type  \
missing_ratio                                                0.0   
num_values                                                     4   
value_ratios   {'First': 82.1, 'Second': 15.9, 'Third': 1.9, ...   

                                                    Company name  
missing_ratio                                                0.0  
num_values                                                    35  
value_ratios   {'Maruti': 19.9, 'Hyundai': 18.5, 'Honda': 10....

### Lưu dữ liệu đã xử lí

In [26]:
raw_df.to_csv("../data/processed/used_cars_data_processed.csv" , index=False)